runtime/lua/vim/iter.lua

```diff
@@ -693,24 +694,54 @@ end
 --- -- 2
 --- it:next()
 --- -- nil
+---
+--- local function pred(x) return x < 2 end
+--- local it2 = vim.iter({ 1, 2, 3, 4 }):take(pred)
+--- it2:next()
+--- -- 1
+--- it2:next()
+--- -- nil
 --- ```
 ---
----@param n integer
+---@generic A
+---
+---@param f integer|fun(a:A, i:integer):boolean Either the number if values to
+---                                             take or a predicate which takes
+---                                             a value and it's index.
 ---@return Iter
-function Iter:take(n)
+function Iter:take(f)
+  if type(f) ~= 'function' then
+    local n = f
+    f = function(_, i)
+      return i <= n
+    end
+  end
+
   local next = self.next
-  local i = 0
+  local i = 1
   self.next = function()
-    if i < n then
+    local val = next(self)
+    if f(val, i) then
       i = i + 1
-      return next(self)
+      return val
```

Comment:
This is not quite right because `next()` can return multiple values. So you could pack it into a table first, and then unpack it. But this is pretty bad for performance. You'll note that the implementations of `filter()` and `map()` have a similar problem and they go to great contortions to address it.

Also we shouldn't pass the index to the function, we don't do that for any other method. If the user wants the index they can call `enumerate()` first which includes the index as one of the iterator values. Staying consistent makes everything more composable.

Simple (but possibly slow) implementation:

```suggestion
    local val = pack(next(self))
    if f(unpack(val)) then
      return unpack(val)
```

For a solution that avoids the `pack` and `unpack` dance you can look at what `map()` does.

Reply:
Thanks for the feedback, I tried a solution without the packing/unpacking similar to `map()` .